File: lib/algorithms/bayes.ts
import type { BayesStep, BayesResult } from '@/lib/types'

export function bayesTheorem(prior: number, likelihood: number, falsePositive: number): BayesResult {
  const pNotDisease = 1 - prior
  const evidence = (likelihood * prior) + (falsePositive * pNotDisease)
  const posterior = (likelihood * prior) / evidence

  const steps: BayesStep[] = [
    { step: 1, description: 'Define Prior Probability P(D)', formula: `P(D) = ${prior}`, value: prior },
    { step: 2, description: 'Define Complement P(¬D)', formula: `P(¬D) = 1 - ${prior} = ${pNotDisease.toFixed(6)}`, value: pNotDisease },
    { step: 3, description: 'Likelihood P(+|D) — True Positive Rate', formula: `P(+|D) = ${likelihood}`, value: likelihood },
    { step: 4, description: 'False Positive Rate P(+|¬D)', formula: `P(+|¬D) = ${falsePositive}`, value: falsePositive },
    { step: 5, description: 'Total Evidence P(+)', formula: `P(+) = P(+|D)·P(D) + P(+|¬D)·P(¬D) = ${evidence.toFixed(6)}`, value: evidence },
    { step: 6, description: 'Apply Bayes Theorem → Posterior', formula: `P(D|+) = P(+|D)·P(D) / P(+) = ${posterior.toFixed(6)}`, value: posterior },
  ]

  return { prior, likelihood, evidence, posterior, steps }
}

export function medicalDiagnosis(
  disease: string, prevalence: number, sensitivity: number,
  falsePositiveRate: number, testResult: 'positive' | 'negative'
): BayesResult {
  if (testResult === 'positive') {
    return bayesTheorem(prevalence, sensitivity, falsePositiveRate)
  }
  const fnr = 1 - sensitivity
  const tnr = 1 - falsePositiveRate
  const pNegEvidence = (fnr * prevalence) + (tnr * (1 - prevalence))
  const posterior = (fnr * prevalence) / pNegEvidence

  const steps: BayesStep[] = [
    { step: 1, description: `Prior: Prevalence of ${disease}`, formula: `P(D) = ${prevalence}`, value: prevalence },
    { step: 2, description: 'False Negative Rate P(−|D)', formula: `P(−|D) = 1 − ${sensitivity} = ${fnr.toFixed(6)}`, value: fnr },
    { step: 3, description: 'True Negative Rate P(−|¬D)', formula: `P(−|¬D) = 1 − ${falsePositiveRate} = ${tnr.toFixed(6)}`, value: tnr },
    { step: 4, description: 'Total Evidence P(−)', formula: `P(−) = ${pNegEvidence.toFixed(6)}`, value: pNegEvidence },
    { step: 5, description: 'Posterior P(D|−)', formula: `P(D|−) = P(−|D)·P(D) / P(−) = ${posterior.toFixed(6)}`, value: posterior },
  ]

  return { prior: prevalence, likelihood: fnr, evidence: pNegEvidence, posterior, steps }
}

export interface SensorInput { name: string; reading: number; reliability: number }
export interface SensorResult { combined: number; confidence: number; steps: BayesStep[] }

export function sensorFusion(sensors: SensorInput[]): SensorResult {
  const steps: BayesStep[] = []
  let combined = 0; let totalWeight = 0

  sensors.forEach((s, i) => {
    const w = s.reliability
    combined += s.reading * w
    totalWeight += w
    steps.push({
      step: i + 1,
      description: `${s.name}: reading=${s.reading.toFixed(2)}, reliability=${s.reliability.toFixed(2)}`,
      formula: `weighted += ${s.reading.toFixed(2)} × ${s.reliability.toFixed(2)} = ${(s.reading * w).toFixed(4)}`,
      value: s.reading * w,
    })
  })

  const result = combined / totalWeight
  const confidence = sensors.reduce((acc, s) => acc * s.reliability, 1)
  const jointConf = 1 - (1 - confidence)

  steps.push({
    step: sensors.length + 1,
    description: 'Weighted average fusion',
    formula: `combined = ${combined.toFixed(4)} / ${totalWeight.toFixed(2)} = ${result.toFixed(4)}`,
    value: result,
  })
  steps.push({
    step: sensors.length + 2,
    description: 'Joint confidence estimate',
    formula: `confidence = 1 − Π(1 − rᵢ) ≈ ${jointConf.toFixed(4)}`,
    value: jointConf,
  })

  return { combined: result, confidence: jointConf, steps }
}

File: backend/uncertainty/bayes.py
import time

def build_bayes_step(step_idx, description, formula, value):
    return {
        "step": step_idx,
        "description": description,
        "formula": formula,
        "value": value
    }

def medical_diagnosis(disease_name, prevalence, sensitivity, fpr, test_result):
    steps = []
    step_idx = 1
    
    prior = prevalence
    steps.append(build_bayes_step(step_idx, "Define Prior Probability", "P(Disease)", prior))
    step_idx += 1
    
    if test_result == 'positive':
        likelihood = sensitivity
        steps.append(build_bayes_step(step_idx, "Likelihood of + given Disease", "P(+|Disease) = Sensitivity", likelihood))
        step_idx += 1
        
        false_positive_prob = fpr
        evidence = (prior * likelihood) + ((1 - prior) * false_positive_prob)
        steps.append(build_bayes_step(step_idx, "Calculate Total Evidence for +", "P(+) = P(+|D)P(D) + P(+|~D)P(~D)", evidence))
        step_idx += 1
        
        posterior = (prior * likelihood) / evidence
        steps.append(build_bayes_step(step_idx, "Calculate Posterior (Bayes' Rule)", "P(D|+) = [P(+|D) * P(D)] / P(+)", posterior))
        
    else:
        likelihood = 1 - sensitivity
        steps.append(build_bayes_step(step_idx, "Likelihood of - given Disease", "P(-|Disease) = 1 - Sensitivity", likelihood))
        step_idx += 1
        
        true_negative_prob = 1 - fpr
        evidence = (prior * likelihood) + ((1 - prior) * true_negative_prob)
        steps.append(build_bayes_step(step_idx, "Calculate Total Evidence for -", "P(-) = P(-|D)P(D) + P(-|~D)P(~D)", evidence))
        step_idx += 1
        
        posterior = (prior * likelihood) / evidence
        steps.append(build_bayes_step(step_idx, "Calculate Posterior (Bayes' Rule)", "P(D|-) = [P(-|D) * P(D)] / P(-)", posterior))

    return {
        "prior": prior,
        "likelihood": likelihood,
        "evidence": evidence,
        "posterior": posterior,
        "steps": steps
    }

def sensor_fusion(sensors):
    steps = []
    step_idx = 1
    
    prior_odds = 1.0 
    current_odds = prior_odds
    
    steps.append(build_bayes_step(step_idx, "Initial Prior Odds", "O(E) = P(E)/P(~E)", current_odds))
    step_idx += 1
    
    combined_reading = 0
    total_weight = 0
    
    for s in sensors:
        reading = s.get('reading', 0)
        rel = s.get('reliability', 0.5)
        name = s.get('name', 'Sensor')
        
        likelihood_ratio = (reading * rel + (1 - reading) * (1 - rel)) / ((1 - reading) * rel + reading * (1 - rel))
        
        steps.append(build_bayes_step(step_idx, f"{name} Likelihood Ratio", "LR", likelihood_ratio))
        step_idx += 1
        
        current_odds *= likelihood_ratio
        
        steps.append(build_bayes_step(step_idx, f"Update Odds with {name}", "O(E|S) = O(E) * LR", current_odds))
        step_idx += 1
        
        combined_reading += reading * rel
        total_weight += rel

    final_prob = current_odds / (1 + current_odds)
    steps.append(build_bayes_step(step_idx, "Convert Odds to Probability", "P = O / (1 + O)", final_prob))
    step_idx += 1
    
    combined_reading = combined_reading / total_weight if total_weight > 0 else 0
    
    return {
        "combined": combined_reading,
        "confidence": final_prob,
        "steps": steps
    }
